src/WeakRefStrings.jl

```diff
@@ -75,19 +74,45 @@ Base.convert(::Type{String}, x::WeakRefString) = convert(String, string(x))
 Base.String(x::WeakRefString) = string(x)
 Base.Symbol(x::WeakRefString{UInt8}) = ccall(:jl_symbol_n, Ref{Symbol}, (Ptr{UInt8}, Int), x.ptr, x.len)
 
-struct WeakRefStringArray{T, N} <: AbstractArray{Union{String, Missing}, N}
+"""
+A [`WeakRefString`](@ref) container.
+Holds the "strong" references to the data pointed by its strings, ensuring that
+the referenced memory blocks stay valid during `WeakRefStringArray` lifetime.
+
+Upon indexing an elemnt in a `WeakRefStringArray`, the underlying `WeakRefString` is converted to a proper
+Julia `String` type by copying the memory; this ensures safe string processing in the general case. If additional
+optimizations are desired, the direct `WeakRefString` elements can be accessed by indexing `A.elements`, where
+`A` is a `WeakRefStringArray`.
+"""
+struct WeakRefStringArray{T, N} <: AbstractArray{T, N}
     data::Vector{Any}
     elements::Array{T, N}
+
+    function WeakRefStringArray(data::Vector{Any}, A::Array{WeakRefString{T}, N}) where {T, N}
+        new{WeakRefString{T}, N}(data, A)
+    end
+
+    function WeakRefStringArray(data::Vector{Any}, A::Array{Union{WeakRefString{T}, Missing}, N}) where {T, N}
+        new{Union{WeakRefString{T}, Missing}, N}(data, A)
+    end
 end
 
-WeakRefStringArray(data::Vector{UInt8}, ::Type{T}, rows::Integer) where {T <: WeakRefString} = WeakRefStringArray(Any[data], zeros(T, rows))
-WeakRefStringArray(data::Vector{UInt8}, ::Type{Union{Missing, T}}, rows::Integer) where {T} = WeakRefStringArray(Any[data], Vector{Union{Missing, T}}(rows))
-WeakRefStringArray(data::Vector{UInt8}, A::Array{T}) where {T <: Union{WeakRefString, Missing}} = WeakRefStringArray(Any[data], A)
+init(::Type{T}, rows) where {T} = fill(zero(T), rows)
+if !isdefined(Base, :uninitialized)
+    struct Uninitialized end
+    const uninitialized = Uninitialized()
+    Vector{T}(::Uninitialized, rows) where {T} = Vector{T}(rows)
+end
+init(::Type{Union{Missing, T}}, rows) where {T} = Vector{Union{Missing, T}}(uninitialized, rows)
+WeakRefStringArray(data, ::Type{T}, rows::Integer) where {T} = WeakRefStringArray(Any[data], init(T, rows))
+WeakRefStringArray(data, A::Array{T}) where {T <: Union{WeakRefString, Missing}} = WeakRefStringArray(Any[data], A)
 
 wk(w::WeakRefString) = string(w)
 wk(::Missing) = missing
 
 Base.size(A::WeakRefStringArray) = size(A.elements)
+Base.eltype(A::WeakRefStringArray{T}) where {T <: WeakRefString} = String
```

Comment:
These shouldn't be needed if `WeakRefStringArray` correctly inherited from `AbstractArray{String}` or `AbstractArray{Union{String, Missing}}`.